app/uk/gov/hmrc/gform/summary/FormComponentSummaryRenderer.scala

```diff
@@ -0,0 +1,924 @@
+
+    val hasErrors = formFieldValidationResult.isNotOk
+
+    val keyClasses = getKeyClasses(hasErrors)
+
+    val label = group.repeatLabel.map(_.value()).getOrElse(fcrd.label(formComponent))
+
+    val visuallyHiddenText = Some(label)
+
+    formComponent.presentationHint match {
+      case Some(hints) if hints.contains(SummariseGroupAsGrid) =>
+        val formFieldValidationResults: List[FormFieldValidationResult] =
+          fcrd.sort(group.fields.filter(fcrd.isVisible)).map(validationResult.apply)
+
+        val errorResults = formFieldValidationResults.filter(_.isNotOk)
+
+        val value =
+          errorResults.headOption match {
+            case None =>
+              formFieldValidationResults.map(ffvr => s"${TextFormatter.formatText(ffvr, envelope)}<br>").mkString
+            case Some(formFieldValidationResult) =>
+              val errors = checkErrors(formComponent, formFieldValidationResult)
+              errors.mkString(" ")
+          }
+
+        if (value.nonEmpty) {
+          List(
+            summaryListRow(
+              label,
+              value,
+              visuallyHiddenText,
+              keyClasses,
+              "",
+              "",
+              if (formComponent.onlyShowOnSummary)
+                Nil
+              else
+                List(
+                  (
+                    uk.gov.hmrc.gform.gform.routes.FormController
+                      .form(
+                        formTemplateId,
+                        maybeAccessCode,
+                        sectionNumber,
+                        sectionTitle4Ga,
+                        SuppressErrors.Yes,
+                        FastForward.Yes),
+                    if (formComponent.editable) messages("summary.change") else messages("summary.view")))
+            )
+          )
+
+        } else List(SummaryListRow())
+
+      case _ =>
+        val rows = fcrd.sort(group.fields.filter(fcrd.isVisible)).flatMap { formComponent =>
+          summaryListRows(
+            formComponent,
+            formTemplateId,
+            formModelVisibilityOptics,
+            maybeAccessCode,
+            sectionNumber,
+            sectionTitle4Ga,
+            obligations,
+            validationResult,
+            envelope
+          )
+        }
+
+        val label = fcrd.label(formComponent)
+        if (label.nonEmpty && formComponent.modelComponentId.maybeIndex.fold(false)(_ === 1)) {
+          val customKeyClasses = "summary-group-label"
+
+          summaryListRow(label, "", None, customKeyClasses, "", "", Nil) :: rows
+        } else rows
+    }
+  }
+}
+
+sealed trait RenderType
+trait InstructionRender extends RenderType
+trait SummaryRender extends RenderType
+
+sealed trait FormComponentRenderDetails[T <: RenderType] {
+  def isVisible(formComponent: FormComponent): Boolean
+  def label(formComponent: FormComponent)(implicit lise: SmartStringEvaluator): String
+  def sort(fields: List[FormComponent]): List[FormComponent]
```

Comment:
Cannot we merge `isVisible` and `sort` into single method? They seems to be used exclusively in combination.

Reply:
I thought so too, but the name would be a bit wierd. sortAndGetRenderables?

Reply:
That `sort` should not be even exposed in the API. It is just detail in `InstructionRender` instance.
I think it would be best to have 

```
sealed trait FormComponentRenderDetails[T <: RenderType] {
  def prepareRenderables(fields: List[FormComponent]): List[FormComponent]
  def label(formComponent: FormComponent)(implicit lise: SmartStringEvaluator): String
}
```

Reply:
Excellent suggestion!